### PR TITLE
Adjustments to #387 (events)

### PR DESCRIPTION
**Pull request by patrick-kidger**
This PR goes through and organises some of the event-handling code. The full list of changes is:

- Semantic change: boolean events now trigger when they become truthy (before they occurred when they swap being falsy<->truthy). Note that this required twiddling around a few things as previously it was impossible for an event to trigger on the first step; now they can.
- Semantic change: event functions now have the signature `(t, y, args *, terms, solver, **etc)` for consistency with vector fields and with `SaveAt(fn=...)`.
- Feature: now backward-compatible with the old discrete terminating events.
- Feature: added `diffrax.steady_state_event`.
- Bugfix: the final `t` and `y` from an event are now saved in the correct index of `ts` and `ys`, rather than just always being saved at index `-1`.
- Bugfix: at one point `args` referred to the `args` coming from a root find rather than the overall `diffeqsolve`.
- Bugfix: the current `state.tprev` was used instead of the previous state's `tnext`. (These are usually but not always the same -- in particular when around jumps.)
- Bugfix: added some checks when the condition function of an event does not return a bool/float scalar.
- Performance: includes a fastpath for skipping the rootfind if no events are triggered.
- Performance: now avoiding tracing for the shape of `dense_info` twice when using adaptive step size controllers alongside events.
- Performance: avoided quadratic loop for figuring out what was the first event to trigger.
- Chore: added support for the possibility of the final root find (for the time of the event) failing.
- Chore: removed some dead code (`_bool_event_gradient`).
- Chore: removed references in the docs to the old `discrete_terminating_event`.

---

In addition, some drive-bys:

- Fixed warnings about pending deprecations `jnp.clip(..., a_min=..., a_max=...)`.
- Had `aux_stats` (in `_integrate.py`) forward to the overall output statistics. In practice this is empty but it's worth doing for the future.

---

Before we can merge this into Diffrax's main branch, there are still some things that need doing -- @cholberg are you able to handle these?

- Add a docstring for `Event`;
- Add an example to `events.md` demonstrating how to use `Event`;
- Update the existing steady-state example to use the new syntax;
- Add several tests:
    - For `steady_state_event`;
    - That things behave correctly when a boolean event stops us before we even perform a single step;
    - Events in combination with adaptive stepping;
    - Events in combination with `vmap` (down various different parameters: `t`, `y`, and even the definition of the event itself!)
    - Events in combination with both adaptive stepping and vmap;
    - Events that are malformed from the user: that we error out appropriately when provided events that return non-scalars, or that return integers, or that return complex numbers.
    - When using more complicated choice of `SaveAt`. For example the previous implementation saved results into index `-1`, which is incorrect in general (e.g. if we have `SaveAt(steps=True)` then we allocate a buffer of length `max_steps`, and only expect to partially fill it).

---

I've tested that this passes all of `test_event.py` (including all the old ones for discrete terminating events), but haven't checked over every test in Diffrax -- it's possible that I've missed something elsewhere.

### CODE REVIEW

**Review by cholberg — APPROVED** [review]
All looks good to me, approved! One question about steady state events: As of right now, you would have to choose between using a `steady_state_event` or defining your own `Event` instance, but in theory is it not possible to treat both of these at once as part of  the same `Event`? I am guessing you are treating steady state events seperately, because there is a seperate adjoint method attached to them?

On the rest: Yes, I got it. It might also be a good idea to add a simple test for repeated events to ensure that the gradients behave correctly when chaining multiple `diffeqsolve`.

**Comment by patrick-kidger** [comment]
That is, you'd like `steady_state_,event` to return just the condition function, not the `Event` object itself? If so then sure, that change seems reasonable.
